examples/flava/native/train.py

```diff
@@ -0,0 +1,316 @@
+            print0(f"after DDP: {torch.cuda.memory_allocated()/1024**3:.3} GB")
+        elif strategy == "fsdp":
+            model = FSDP(
+                model,
+                mixed_precision=MixedPrecision(
+                    # param_dtype=torch.bfloat16  doesn't work
+                   reduce_dtype=torch.bfloat16,
+                    # buffer_dtype=torch.bfloat16, checkpoint not supported
+                ),
+                auto_wrap_policy=partial(
+                    transformer_auto_wrap_policy,
+                    transformer_layer_cls={TransformerEncoderLayer},
+                ),
+            )
+            print0(f"after FSDP {torch.cuda.memory_allocated()/1024**3:.3} GB")
+
+        else:
+            raise ValueError(f"unknown strategy: {strategy}")
+
+        print0(
+            f"after {strategy} model parameters: {get_model_parameters(model):,}, "
+            f"size: {get_model_size_gb(model):.3} GB"
+        )
+
+        return model
+
+    def calculate_loss(
+        self, output: FLAVAPretrainingLossOutput, validation=False
+    ) -> torch.Tensor:
+        losses = output.losses
+
+        total_loss = 0
+        for key in losses:
+            if losses[key] is not None:
+                total_loss += losses[key]
+                loss_reduce = losses[key].detach()
+                dist.reduce(loss_reduce, dst=0)
+                if validation:
+                    mode = "validation"
+                else:
+                    mode = "train"
+                    self.log(
+                        f"{mode}/losses/{key}",
+                        loss_reduce.item() / dist.get_world_size(),
+                    )
+
+        return total_loss
+
+    def preprocess_data(self, data: Dict[str, Any]):
+        data = self.datamodule.on_before_batch_transfer(data, None)
+        data = move_to_device(data, self.device)
+        return self.datamodule.on_after_batch_transfer(data, None)
+
+    def train(self) -> None:
+        print0(OmegaConf.to_container(self.config.training))
+        self.model = self.create_model()
+        model = self.model
+
+        optimizer, scheduler = get_optimizer(
+            model,
+            learning_rate=self.config.training.get("learning_rate"),
+            adam_eps=self.config.training.get("adam_eps"),
+            adam_weight_decay=self.config.training.get("adam_weight_decay"),
+            adam_betas=self.config.training.get("adam_betas"),
+            warmup_steps=self.config.training.get("warmup_steps"),
+            max_steps=self.config.training.get("max_steps"),
+            use_bf16=self.config.training.get("enable_bf16"),
+        )
+
+        while True:
+            t0 = time.time()
+            self.epochs += 1
+            dataloader = self.datamodule.train_dataloader()
+            dataloader.set_epoch(self.epochs)
+
+            for i, data in enumerate(dataloader):
+                torch.cuda.reset_peak_memory_stats()
+
+                self.steps += 1
```

Comment:
do u need to sync this?

Reply:
hm, not that i'm aware. why would they be synced?